### PR TITLE
create /template route from unused faq route

### LINKED ISSUE #113: convert /faq into /template route
We need a proper template for our standard route. With a server fetch in page.tsx and a page.client.tsx reflecting some of our fleshed out routes like /commodities. The goal is to provide a scaffolding that can be copied when new routes are built. Also, a template route would allow us to test new features on a route closely resembling what's common in the app. Features like: error handling, file structure updates, fetch changes, etc.

The page.client.tsx should include some sort of fetch call, even it's as simple as a button that makes a call then renders the response. Bonus points if the UI includes a simple form and table like the /commodities route. Many of our routes will include this, so having them here will save time in the future.

### PR DESCRIPTION
**Pull request by aslink87**
establishes a template route to scaffold future routes
includes:

1. server side fetch with error handling
2. page layout reflecting current app styles to include a form component
3. client side fetch prompted by form submission with error handling

Discussion points:

1. The server side data call redirects to the nearest error page on error. This makes sense as that data should be vital to the functionality of the page.
2. However, the client side call does not redirect to an error page. Instead it renders an error alert under the form. My thought was to not break the page if the form submission failed. Is that fine with you guys, or would you rather it present an error page?

Screenshot 2023-09-15 at 14-13-52 EDPN Template

4. Chris, you've made far more pages than anyone, so I'd appreciate if you could check to see if there's something that can be refactored here to simplify the code. Or if there's something I could've done better to make this template route easier to understand if a new dev wanted to use it.
5. I included the form opacity to match the /commodities route if a background image were to be added later

### CODE REVIEW

**Comment by aslink87** [comment]
screenshot-Desktop-Chrome-1920x1080-dark

screenshot-Desktop-Chrome-1280x1024-light

screenshot-Desktop-Chrome-768x1024-dark

screenshot-Desktop-Chrome-480x720-light

File: app/_components/utility/pageHeading.tsx
Comment (aslink87) [inline app/_components/utility/pageHeading.tsx:21]: @chrismou would you prefer this subheading to be optional? I noticed you don't always use it.

Comment (nitishdhar) [reply app/_components/utility/pageHeading.tsx:21]: My pick would be to keep it optional. But confirm with @chrismou

File: app/template/page.client.tsx
Comment (nitishdhar) [inline app/template/page.client.tsx]: If we can keep some comments in the template file where it clarifies what is the purpose of that component e.g., here I believe the dev has to replace or remove this depending on the data type of the response. Totally up for debate if you think we should have comments in the file where relevant.

File: app/template/page.client.tsx
Comment (nitishdhar) [inline app/template/page.client.tsx]: Nice :) this is the first thing we do on a page which fetches data

File: app/template/page.tsx
Comment (nitishdhar) [inline app/template/page.tsx]: I understand what you are doing here and correct me if my assumption is wrong. Any failure here will show the closest error boundary. This is fine for now and I am not expecting any change in this PR, but ideally we want to create a typed schema for data that's sent to components as props. I haven't thought of the exact implementation but something on the lines of - if error on server side, set error and assign it to the component prop which can take success & error states so we can still show this alert of error in a localized section of the page. Let me think about it and create a discussion as an issue. Let me know your thoughts though.

**Review by nitishdhar — APPROVED** [review]
LGTM, pending discussions in the review are not blockers for me. Thanks for setting this up @aslink87

File: app/_components/utility/pageHeading.tsx
Comment (chrismou) [reply app/_components/utility/pageHeading.tsx:21]: I could shoehorn subheadings on to all pages, but I don't think they all need them so would probably say optional?

File: app/template/api/index.tsx
Comment (chrismou) [inline app/template/api/index.tsx]: I originally wrote some comments here they're better placed here https://github.com/ed-pilots-network/frontend/pull/119/files#diff-228a314c046a3a9b4c3e7ff6bdf0dc14c18826b9e0eea866b4def74d54b2e1dfR10 and both sort of tie in with each other

File: app/template/page.client.tsx
Comment (chrismou) [inline app/template/page.client.tsx:79]: Can this be broken down any further? I noticed you did a PageHeader component which is awesome, really simplifies keeping layouts consistent, so could this potentially be broken down into a ContentArea component or something?

File: app/template/page.tsx
Comment (chrismou) [inline app/template/page.tsx:13]: With this API call being used to get commodity data for a dropdown on about 4-5 pages and will always be the same, does it make sense to move it all into a hook or a faux hook? I'm just thinking that this function will be copied pretty much verbatim for each page, so maybe we could simplify it further by having a useGetCommodities() we can just drop onto a page (or a non hook function if we think it might get used outside of a tsx component)

I appreciate it's only an example, but you highlighted the point quite well in this example where you've made 2 identical queries client and server side. Which I get this is niche, it could be nice to be able to just call `getCommodities(paramsObject)` and have a reusable hook/function/whatever we can call so we don't need to start hunting around for urls, cache times other developers have used, etc

Or maybe go more generic with a getApiRequest function where you specify the endpoint and the params?

I've probably not spent enough time thinking about it, my line of thinking is simply that every time I created a page with a commodity drop down, I'd need to search the codebase for the URL, the cache time, and probably copypaste a bunch of code, or if someone suggests increasing cache time we would need to change it in several place - which just shouts to me that it should be abstracted. 

Perhaps it just needs a bit of further discussion, or maybe we just accept that it's not enough of a hassle to warrant extra work and we can just live with it? Unsure tbh. Thoughts?

**Review by chrismou — CHANGES_REQUESTED** [review]

File: app/template/page.tsx
Comment (nitishdhar) [reply app/template/page.tsx:13]: @chrismou I have a working change locally that will implement swr based hooks (https://github.com/ed-pilots-network/frontend/issues/105), with all the open PRs, I have been holding on to that change and waiting for this one to be merged as I saw @aslink87 has moved the API call out and will make what I am doing even simpler.

Hook Example - app/_hooks/_api
```javascript
import useSWR from 'swr';
import { ICommodity } from '@/app/_types';

export default function useCommodity(name: string) {
  const { data, error, isLoading } = useSWR(
    `${process.env.NEXT_PUBLIC_MOCK_API_URL}/api/v1/trade/commodity/${name}`,
  );

  return {
    commodity: data as ICommodity,
    isLoading,
    isError: error as Error,
  };
}
```

Using the hook

```javascript
  import useCommodity from '@/hooks/api/useCommodity';
  const { commodity, isLoading, isError } = useCommodity('Tritium');

  const renderCommodityData = () => {
    if (isLoading) {
      return <Spinner />;
    }
    if (isError) {
      return (
        <Alert status="error">
          <AlertIcon />
          Failed to fetch commodity data from staging server on client side
        </Alert>
      );
    }
    if (commodity) {
      return <div>Commodity Name: {commodity.commodityName}</div>;
    }
```

File: app/template/page.tsx
Comment (aslink87) [reply app/template/page.tsx]: Correct about the error. This will go to the root error page you created recently.

To have what you're saying be useful, that is to pass the error along and provide it as feedback on the page, the error would have to be non-vital to the page's intent. There may be cases where that occurs, but atm I think all of our server fetches populate form fields that are required. Therefore, they are blockers to the form. But as the app grows, we may be happy to do what you are suggesting now, so I'm fine with that route too.

File: app/template/page.tsx
Comment (aslink87) [reply app/template/page.tsx:13]: Nextjs recommends swr for client-side fetching [here](https://nextjs.org/docs/pages/building-your-application/data-fetching/client-side#client-side-data-fetching-with-swr).

If that's the case, then we would continue fetching the server data that's needed in the pages within the page.tsx similar to how I've done it here. We could use swr for the client side form submissions.

Comment (aslink87) [reply app/template/page.tsx:13]: I like where @chrismou was heading in his comment. Correct me if i'm wrong but all our server fetches are for arrays of data used in form elements. I don't see that changing. So we could look into extracting that server call and calling it in each page.tsx where it is needed. `getFormProps(url)`.

Then if we go with swr, we could have a universal submitForm function to return the form's result.

File: app/template/page.client.tsx
Comment (aslink87) [reply app/template/page.client.tsx:79]: It certainly could. So this is returning the results of the form. The first question is, are those results going to be displayed in a universal way across our pages? We may not be able to answer this now. In the commodities route, I kept the results in the route. Looking at how that data is displayed, can it be made generic enough to create a results component that can be reused?

I think this is better saved for another PR. Here, i tried to render the results in a simple manner just so devs can see that the fetch calls worked.

File: app/template/page.tsx
Comment (nitishdhar) [reply app/template/page.tsx]: It's good now for, we anyway need to think more about how we will handle that and if we even need it for a recoverable or non-critical error.

File: app/template/page.tsx
Comment (nitishdhar) [reply app/template/page.tsx:13]: @aslink87 Extracting that server call to where? like as a swr hook? I don't think we can use hooks in server pages, but if you mean we should move it to a common place so other pages can use the same method (not a swr hook), then yeah I am aligned.

**What we all aligned on (correct if anything is wrong)**

- Client side fetches will be made re-usable with swr based hooks and 
- server side can be just functions packed under some server side api utility. Similar to what you have in this change and what I have in the playground [API ](https://github.com/ed-pilots-network/frontend/blob/development/app/playground/_api/index.tsx)

Comment (aslink87) [reply app/template/page.tsx:13]: Tomorrow I'll move the server call to a more universal location. @nitishdhar Do you want to add your SWR implementation right into this branch? would be nice to have a complete picture here before merging into dev.

Comment (nitishdhar) [reply app/template/page.tsx:13]: Will do 👍

File: app/template/page.client.tsx
Comment (aslink87) [reply app/template/page.client.tsx]: done

File: app/template/page.tsx
Comment (aslink87) [reply app/template/page.tsx:13]: extracted api calls to universal functions: f00017268f18037582c76acd74427b16dae83bb9

swr can replace the `getSubmitFormClient` fn from this commit

Comment (nitishdhar) [reply app/template/page.tsx:13]: @aslink87 I have implemented swr in this branch. Couple of things - I noticed you have made the fetch as general calls so I kept the new hook also under the same library file. An alternate would be to move something like `client-api-calls` under `hooks` folder as all client fetches will now be hooks and keep server side calls under `_lib/server-api-calls`. I did it but then realized there's only one for both for now and that might not change unless out fetch patterns change so reverted to just that one library file. Also, replaced the custom response with a specific type with updated comments to replace with relevant resource type - https://github.com/ed-pilots-network/frontend/pull/119/commits/2cd2090e4ce6e5b6f80ee77d0eb43e68fc54d876

**Review by chrismou — APPROVED** [review]

**Pull request merged**
